service/history/ndc/history_replicator.go

```diff
@@ -227,6 +228,196 @@ func (r *HistoryReplicatorImpl) ApplyEvents(
 	return r.doApplyEvents(ctx, task)
 }
 
+func (r *HistoryReplicatorImpl) BackfillHistoryEvents(
+	ctx context.Context,
+	request *shard.BackfillHistoryEventsRequest,
+) error {
+	task, err := newReplicationTaskFromBatch(
+		r.clusterMetadata,
+		r.logger,
+		request.WorkflowKey,
+		request.BaseExecutionInfo,
+		request.VersionHistoryItems,
+		request.Events,
+		request.NewEvents,
+		request.NewRunID,
+		request.VersionedHistory,
+	)
+	if err != nil {
+		return err
+	}
+
+	return r.doApplyBackfillEvents(ctx, task, r.applyBackfillEvents)
+}
+
+func (r *HistoryReplicatorImpl) doApplyBackfillEvents(
+	ctx context.Context,
+	task replicationTask,
+	action func(context.Context, workflow.MutableState, workflow.Context, wcache.ReleaseCacheFunc, replicationTask) error,
+) (retError error) {
+	wfContext, releaseFn, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		task.getNamespaceID(),
+		task.getExecution(),
+		locks.PriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if rec := recover(); rec != nil {
+			releaseFn(errPanic)
+			panic(rec)
+		}
+		releaseFn(retError)
+	}()
+
+	mutableState, err := wfContext.LoadMutableState(ctx, r.shardContext)
+	switch err.(type) {
+	case nil:
+		return action(ctx, mutableState, wfContext, releaseFn, task)
+	case *serviceerror.NotFound:
+		return serviceerrors.NewSyncState(
+			mutableStateMissingMessage,
+			task.getNamespaceID().String(),
+			task.getWorkflowID(),
+			task.getRunID(),
+			task.getVersionedTransition(),
+		)
+	default:
+		return err
+	}
+}
+
+func (r *HistoryReplicatorImpl) applyBackfillEvents(
+	ctx context.Context,
+	mutableState workflow.MutableState,
+	wfContext workflow.Context,
+	releaseFn wcache.ReleaseCacheFunc,
+	task replicationTask,
+) (retError error) {
+
+	versionedTransition := task.getVersionedTransition()
+	if versionedTransition == nil {
+		return &serviceerror.InvalidArgument{
+			Message: "versioned transition is required for backfill task",
+		}
+	}
+
+	if task.getFirstEvent().GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED {
```

Comment:
Does this mean we are not expecting to backfill first event of workflow? If the case, is it more safe to check if the first event id is `1`?

Reply:
I am okay with checking event type, this is more for discussion.
If this check is to make sure backfill is not started with first event, then it makes more sense to check the event id?
If check event type, it means we are making an assumption that first event is always `EVENT_TYPE_WORKFLOW_EXECUTION_STARTED`, it might be the case for now, but could change in the future?

Reply:
Yes there is this assumption in current code and enforced with check [here](https://github.com/temporalio/temporal/blob/e2dfb2e16f8c0bee3a2209737ef7e1df69fc3cc7/service/history/workflow/mutable_state_impl.go#L2148). Prefer to check the event type here to be consistent with existing ndc and ms apply events implementations.